src/coreclr/nativeaot/Runtime/startup.cpp

```diff
@@ -275,6 +304,53 @@ bool DetectCPUFeatures()
 }
 #endif // !USE_PORTABLE_HELPERS
 
+inline
+GSCookie * GetProcessGSCookiePtr() { return  const_cast<GSCookie *>(&s_gsCookie); }
+
+bool InitGSCookie()
+{
+    volatile GSCookie * pGSCookiePtr = GetProcessGSCookiePtr();
+
+    // The GS cookie is stored in a read only data segment
+    uint32_t oldProtection;
+#if TARGET_WINDOWS
+    uint32_t * pOldProtection = &oldProtection;
+#else
+    // PAL layer is unable to extract old protection
+    uint32_t * pOldProtection = NULL;
+    oldProtection = PAGE_READONLY;
+#endif
+    
+    if (!PalVirtualProtect((void*)pGSCookiePtr, sizeof(GSCookie), PAGE_READWRITE, pOldProtection))
+    {
+        return false;
+    }
+
+#if TARGET_WINDOWS
+    // The GSCookie cannot be in a writeable page
+    assert(((oldProtection & (PAGE_READWRITE|PAGE_WRITECOPY|PAGE_EXECUTE_READWRITE|
+                              PAGE_EXECUTE_WRITECOPY|PAGE_WRITECOMBINE)) == 0));
+
+    // Forces VC cookie to be initialized.
+    void * pf = &__security_check_cookie;
```

Comment:
Can we use the security cookie used by C/C++ directly instead of creating our local copy?

Reply:
Would there be an advantage? Right now the advantage is that we use the same thing on both Linux and Window. We just seed it differently and happen to seed it from `__security_cookie`.

We don't own `__security_cookie` or policies around when it's initialized (that's why we need to do the dance with `void * pf = &__security_check_cookie;` above, I assume) so I'm a bit worried about using it directly.

Reply:
Flipping the page protection from RO to RW on page in a mapped binary is suspicious activity. 

On Windows, the real GS cookie is initialized by the OS loader via `IMAGE_LOAD_CONFIG_DIRECTORY` in all recent Windows versions so that the user mode does not need to do this suspicious activity and gets guaranteed good GS cookie. Not sure how it works on Linux.